Improve Pixiv authorization

n-ski opened the pull request:
This PR enables Pixiv authorization through `Auth` property.

The changes include:
- Removed `AccessToken` and `RefreshToken` properties, and `_refreshTime` field. This means access and refresh tokens are no longer publicly visible (not that they should have been visible in the first place IMO).
- `SessionInfo` struct only visible to `Pixiv` was introduced to hold all the abovementioned data in one place, and the data is now stored in the `_sessionInfoTask` field, a task that returns `SessionInfo` value.
- `ABooru`'s `Auth` property was made virtual. `Pixiv` overrides this property to inject its own logic that starts a task that retrieves session info for its methods to use.
- `LoginAsync(string, string)` method was removed completely since it is now superseded by `Auth` property.

Now all the methods that require access tokens in order to work wait for the `_sessionInfoTask` to complete internally, so setting `Auth` immediately followed by a Pixiv API call (or several calls, since they all await on the same task instance) is completely safe. In contrast previously you had to wait until `LoginAsync(string, string)` completes before you could safely perform any Pixiv API call.

Other changes not directly related to the changes above:
- Updated `Pixiv` documentation: added missing exceptions and added remaks about required authorization.
- Implemented `IEquatable` interface for `BooruAuth` class, it compares properties of two `BooruAuth` objects to determine whether they are equal. `Pixiv`'s `Auth` property makes use of this so it doesn't unnecessarily create a new authorization task even if credentials are the same.
- Refactored a bunch of repeated code into separate methods in `Pixiv`.
- Made `UnitTests.Boorus` class' methods synchronous since now there's no asynchronous method calls.

n-ski committed:
Improve Pixiv authorization

- Removed LoginAsync(string, string) method. Instead users need to authorize using Auth property, which updates private Task field responsible for retrieving session info (access and refresh tokens, and expiration date). None of that data is now publicly exposed.

To achieve the above ABooru's Auth property was made virtual and Pixiv overrides this property. Overridden setter will start a task that retrieves session info from Pixiv API and will store the task in a private field.

- Updated Pixiv documentation: added missing exceptions and added remaks about required authorization.

- Implemented IEquatable interface for BooruAuth class, it compares properties of two BooruAuth objects to determine whether they are equal. Pixiv's Auth property makes use of this.

- Refactored a bunch of repeated code into separate methods in Pixiv class.

- Changed favorite API unit tests to make sure booru's Auth property is properly reset to its previous state after the test is over.

n-ski committed:
Make UnitTests.Boorus class' methods synchronous

n-ski committed:
Make Pixiv session info thread-safe

Use simple locks to achieve thread safety.

Initially I planned to use reader/writer with read locks wrapping the whole method bodies (including 'await's) but judging by unit test results that would introduce deadlocks, so simple locks are used for reading/writing the task field itself and not locking for the duration of the task.

n-ski commented on BooruSharp.Others/Pixiv.cs:
```diff
@@ -27,76 +27,40 @@ public Pixiv()
         /// <exception cref="AuthentificationInvalid"/>
         /// <exception cref="HttpRequestException"/>
-        public async Task LoginAsync(string refreshToken)
+        public Task LoginAsync(string refreshToken)
```
Not sure if it's worth keeping this method at all. Authorization through refresh token is now implemented internally and I don't see anyone using their refresh token to log in at all.

Xwilarg replied:
Authentification by refresh token allow someone to just reuse his token, or take it from somewhere else and login in BooruSharp using it

Xwilarg commented:
Tokens were public so you could just use your own, or retrieve the one that BooruSharp got

n-ski commented on BooruSharp.Others/Pixiv.cs:
```diff
@@ -128,108 +92,86 @@ public async Task<byte[]> PreviewToByteArrayAsync(SearchResult result)
         /// <exception cref="AuthentificationInvalid"/>
+        /// <exception cref="AuthentificationRequired"/>
         /// <exception cref="HttpRequestException"/>
+        // TODO: does this have to be public?
         public async Task CheckUpdateTokenAsync()
```
Same here. I don't think this has to be explicitly called by the user.

Xwilarg replied:
Iirc this was used when you provide your own token

n-ski replied:
But since all the other methods call this method to validate the auth info as well as update the tokens, does it make sense to keep `CheckUpdateTokenAsync` `public`?

Xwilarg replied:
Speaking of that, looks like the favorite functions don't do that, that may be a mistake from me

KoB-Kirito commented:
You need to be able to provide a refresh token. If you regain it everytime, you get send mails from pixiv everytime. I don‘t see a reasonable way to store those persistent on library level.

Edit: So you would also need to get it, to store it..

KoB-Kirito reviewed (commented):
Pixiv may actually block your account if you login without refresh token everytime...

n-ski commented:
>Pixiv may actually block your account if you login without refresh token everytime...

I see, I'll restore `AccessToken` and `RefreshToken` properties and keep `LoginAsync(string)` as is.

n-ski committed:
Address PR review comments

- Changed CheckUpdateTokenAsync visibility to 'private'
- Restored AccessToken and RefreshToken properties, they are backed by session info task field
